lib/puppet/util/windows/service.rb

```diff
@@ -0,0 +1,437 @@
+    #   the service
+    def self.open_service(service_name)
+      scm = OpenSCManagerW(nil, nil, SC_MANAGER_ALL_ACCESS)
+      raise Puppet::Util::Windows::Error.new(_("Failed to open a handle to the service control manager")) if scm == FFI::Pointer::NULL_HANDLE
+      service = OpenServiceW(scm, service_name, SC_MANAGER_ALL_ACCESS)
+      raise Puppet::Util::Windows::Error.new(_("Failed to open a handle to the service")) if service == FFI::Pointer::NULL_HANDLE
+      yield(service)
+    ensure
+      CloseServiceHandle(scm) if scm
+      CloseServiceHandle(service) if service
+    end
+
+    # @api private
+    def self.process_wait_hint(wait_hint)
+      # The first divisor begs a little explanation: the value of
+      # :dwWaitHint will return in milliseconds, but ruby sleep
+      # takes seconds. So we need at least / 1000. After that, the
+      # suggested operation is to wait 1/10 of the actual wait hint
+      # between querying the service. So we get wait_hint / 10000
+      wait_time = wait_hint / 10000;
+      wait_time = 1 if wait_time < 1
+      wait_time = 10 if wait_time > 10
+      wait_time
+    end
+
+    # @api private
+    # perform QueryServiceStatusEx on a windows service and return the
+    # result
+    #
+    # @param [:handle] service handle of the service to query
+    # @return [SERVICE_STATUS_PROCESS struct] the result of the query
+    def self.query_status(service)
+      status = SERVICE_STATUS_PROCESS.new
+      success = false
+      FFI::MemoryPointer.new(:lpword) do |bytes_pointer|
+        success = QueryServiceStatusEx(
+          service,
+          SC_STATUS_PROCESS_INFO,
+          status,
+          status.size,
+          bytes_pointer
+        )
+      end
+      unless success
+        raise Puppet::Util::Windows::Error.new(_("Service query failed"))
+      end
+      status
+    end
+
+    # @api private
+    # perform QueryServiceConfigW on a windows service and return the
+    # result
+    #
+    # @param [:handle] service handle of the service to query
+    # @return [QUERY_SERVICE_CONFIGW struct] the result of the query
+    def self.query_config(service)
+      config = QUERY_SERVICE_CONFIGW.new
+      success = false
+      FFI::MemoryPointer.new(:lpword) do |bytes_pointer|
+        success = QueryServiceConfigW(
+          service,
+          config,
+          config.size,
+          bytes_pointer
+        )
+      end
+      unless success
+        raise Puppet::Util::Windows::Error.new(_("Service query failed"))
+      end
+      config
+    end
+
+    # @api private
+    # waits for a windows service to report SERVICE_STOPPED if it
+    # is in SERVICE_STOPPED_PENDING
+    #
+    # @param [:handle] service handle to the service to wait on
+    # @return [bool] 'true' once the service is reporting SERVICE_STOPPED,
+    #   'false' if the service was not waiting to stop or queries failed
+    def self.wait_for_service_pending(service, pending_state, complete_state)
```

Comment:
Maybe rename to `wait_for_pending_transition(service, pending_state, final_state)`? Also, I think this would be better to read if it was written as:
```
loop do
  state = service_state(service)
  return if state == final_state
  unless state == pending_state
    // Raise error since it's an unknown state ?
  end
  // wait-hint stuff here
end
```